File: T-Bank/Algorithms-and-Data-Structures-2025/4sprint-SolvingProblemMethodsPart1/contest/H-Межрегиональная_олимпиада/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(NlogN)
    Memory Complexity O(1)
    """
    n, c = map(int, sys.stdin.readline().rstrip().split())
    segments: list[tuple[int, int, int]] = []
    for i in range(1, n + 1):
        start, duration = map(int, sys.stdin.readline().rstrip().split())
        segments.append((start, start + duration, i))
    segments.sort(key=lambda task: (task[1], task[0]))
    outcome: int = 0
    prev_end: int = 0
    task_id: int = 0
    path: list[int] = []
    for segment in segments:
        if segment[0] >= prev_end:
            if prev_end:
                path.append(task_id)
                outcome += c
            prev_end = segment[1]
            task_id = segment[2]
    outcome += c 
    path.append(task_id)
    sys.stdout.write('{}\n{}\n'.format(outcome, len(path)))
    sys.stdout.write(' '.join(str(task) for task in path) + '\n')


if __name__ == '__main__':
    solution()

File: T-Bank/Algorithms-and-Data-Structures-2025/4sprint-SolvingProblemMethodsPart1/contest/H-Межрегиональная_олимпиада/solution.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <tuple>


bool compareSegments(const std::tuple<long long, long long, int>& x, const std::tuple<long long, long long, int>& y) {
    if (std::get<1>(x) != std::get<1>(y)) return std::get<1>(x) < std::get<1>(y);
    return std::get<0>(x) < std::get<0>(y);
}


void solution() {
    /*
    Time Complexity O(NlogN)
    Memory Complexity O(1)
    */
    int n, c;
    std::scanf("%d %d", &n, &c);
    std::vector<std::tuple<long long, long long, int>> segments;
    for (int i = 1; i <= n; ++i) {
        long long start, duration;
        std::scanf("%lld %lld", &start, &duration);
        segments.push_back({start, start + duration, i});
    }
    std::sort(segments.begin(), segments.end(), compareSegments);
    std::vector<int> path;
    long long outcome = 0;
    long long prevEnd = 0;
    int taskId = 0;
    for (std::tuple<long long, long long, int>& segment : segments) {
        if (std::get<0>(segment) >= prevEnd) {
            if (prevEnd) {
                path.push_back(taskId);
                outcome += (long long)c;
            }
            prevEnd = std::get<1>(segment);
            taskId = std::get<2>(segment);
        }
    }
    outcome += c;
    path.push_back(taskId);
    std::printf("%lld\n%d\n", outcome, (int)path.size());
    for (int i = 0; i < path.size(); ++i) {
        if (i) std::printf(" ");
        std::printf("%d", path[i]);
    }
    if (!path.empty()) std::printf("\n");
}


int main() {
    solution();
}